035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,280 @@
+
+#### Limit Types
+1. Consumed space limit: Triggers if the consumed space of a volume breaches the value. Candidate for deprecation.
+2. Minimum Free Bytes: Triggers if the amount of free space on a volume drops below the configured level. Particularly
+   useful for hard limits as an absolute minimum of free space.
+3. Minimum Free Percentage. Triggers if the amount of free space on the volume drops below the configured proportion of
+   the volumes total capacity.
+
+#### The data source should:
+- Collect local volume usage
+- Publish usage to an internal compacted topic keyed by broker ID.
+- Collect and publish volume usage metrics (using this [message schema](#Message-schema)) on start up and periodically after that.
+- Connect as an internal service using TLS and client certificates to connect to the Replication listener.
+
+#### The data sink should:
+- Connect as an internal service using TLS and client certificates to connect to the Replication listener.
+- On startup and periodically thereafter read the volume usage stats from the topic
+- Use the Kafka Admin client to discover the list of currently active brokers
+
+#### The quota policy should:
+- Should apply the largest delay caused by each disk which breaches the soft limit.
+    - The amount of throttling should be proportional to how much of the difference between the soft and hard limit
+      remains for a given volume. Applying more throttle the closer to the hard limit things are.
+- Should apply a `PAUSE` level throttle to client produce requests if **any** volume in the cluster breaches the hard
+  limit.
+- Ensure that the quota plugin, and other Strimzi services, updates are not throttled.
+- Should take a configured action when it can not determine the state of all known brokers.
+    - The action should be one of `PAUSE` or `OPEN`.
+    - Defaulting to `PAUSE` as the fail-safe option.
+- Apply a freshness check to the volume usage it reads and ignore stale state [3]. The threshold for staleness should be a configuration parameter.
+
+#### Implementation details
+##### Internal API within the quota plugin
+
+To make all this work the quota plugin will need to introduce some new interfaces:
+
+1. A Quota Policy - To encapsulate the logic for evaluating if a volume `complies` or `breaches` a policy
+2. A Quota Factor Source - To calculate the factor in the range `0..100`% to apply to the entries in the quota map
+3. A Data Source - To publish the current volume usage for the broker in question.
+
+One can envisage a model where the Quota policy was externalised to a separate service which calculated the quota factor
+and thus the Quota Factor Source would just provide a wrapper around that external service. For the purposes of this
+proposal all three interfaces would be implemented with the current quota plugin.
```

Comment:
It's not clear to me whether each instance of the plugin has an active instance of the policy, making local decisions, or whether there's a single active instance of the policy for the whole cluster (for example, the consumer group leader)

Reply:
I've updated the text to clarify I envisage each broker operating independently.